From fe8f43befbdc5241d8c0c687d50114f4eb77fcef Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sat, 9 Jun 2018 20:00:41 +0000
Subject: [PATCH] CONFIGURATION-703: Improved handling of xml:space="preserve".

For tags whose value consists only of whitespace this content is now
returned correctly if xml:space is set to "preserve". Thanks to
Pascal Essiembre for the patch.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1833250 13f79535-47bb-0310-9956-ffa450edef68
---
 .../commons/configuration2/XMLConfiguration.java      |  6 +++++-
 .../commons/configuration2/TestXMLConfiguration.java  |  5 ++++-
 src/test/resources/test.xml                           | 11 ++++++-----
 3 files changed, 15 insertions(+), 7 deletions(-)

diff --git a/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java b/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java
index 0eba290888..ae6f01ea3c 100644
--- a/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java
@@ -630,7 +630,11 @@ else if (w3cNode instanceof Text)
             }
         }
 
-        boolean childrenFlag = hasChildren || attributes.size() > 1;
+        boolean childrenFlag = false;
+        if (hasChildren || trimFlag)
+        {
+            childrenFlag = hasChildren || attributes.size() > 1;
+        }
         String text = determineValue(buffer.toString(), childrenFlag, trimFlag);
         if (text.length() > 0 || (!childrenFlag && level != 0))
         {
diff --git a/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java b/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java
index dec88784eb..dcd4fc0a62 100644
--- a/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java
@@ -1458,7 +1458,10 @@ public void testPreserveSpace()
     @Test
     public void testPreserveSpaceOnElement()
     {
-        assertEquals("Wrong value", " preserved ", conf.getString("spaceElement"));
+        assertEquals("Wrong value spaceElement",
+                " preserved ", conf.getString("spaceElement"));
+        assertEquals("Wrong value of spaceBlankElement",
+                "   ", conf.getString("spaceBlankElement"));
     }
 
     /**
diff --git a/src/test/resources/test.xml b/src/test/resources/test.xml
index 1f52464e36..383db12ae4 100644
--- a/src/test/resources/test.xml
+++ b/src/test/resources/test.xml
@@ -40,7 +40,7 @@ And even longer.
     <test>
         <short>8</short>
     </test>
-	
+
 	<!-- list properties -->
 	<list>
 		<item name="one">one</item>
@@ -54,7 +54,7 @@ And even longer.
 			<item>six</item>
 		</sublist>
 	</list>
-    
+
     <!-- Comma delimited lists (work in elements, but not in attributes) -->
     <split>
       <list1>a,b,c</list1>
@@ -82,19 +82,19 @@ And even longer.
 			<item id="4">four</item>
 		</list>
 	</clear>
-    
+
     <!-- Complex property names -->
     <complexNames>
       <my.elem>Name with dot
         <sub.elem>Another dot</sub.elem>
       </my.elem>
     </complexNames>
-    
+
     <!-- An empty element. This should occur in the configuration with an
          empty string as value.
     -->
     <empty/>
-    
+
     <!-- List nodes with attributes -->
     <attrList>
       <a name="x">ABC</a>
@@ -116,4 +116,5 @@ And even longer.
       <testInvalid xml:space="invalid">     Some other text </testInvalid>
     </space>
     <spaceElement xml:space="preserve"> preserved </spaceElement>
+    <spaceBlankElement xml:space="preserve">   </spaceBlankElement>
 </testconfig>
